src/ImageSharp/Processing/Processors/Quantization/PaletteQuantizer{TPixel}.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Runtime.CompilerServices;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.ImageSharp.Processing.Processors.Dithering;
+
+namespace SixLabors.ImageSharp.Processing.Processors.Quantization
+{
+    /// <summary>
+    /// A generic palette quantizer.
+    /// </summary>
+    /// <typeparam name="TPixel">The pixel format.</typeparam>
+    public class PaletteQuantizer<TPixel> : IQuantizer
+            where TPixel : struct, IPixel<TPixel>
+    {
+        private readonly TPixel[] palette;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PaletteQuantizer{TPixel}"/> class.
+        /// </summary>
+        /// <param name="palette">The color palette to use.</param>
+        public PaletteQuantizer(TPixel[] palette)
+             : this(palette, true)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PaletteQuantizer{TPixel}"/> class.
+        /// </summary>
+        /// <param name="palette">The color palette to use.</param>
+        /// <param name="dither">Whether to apply dithering to the output image</param>
+        public PaletteQuantizer(TPixel[] palette, bool dither)
+            : this(palette, GetDiffuser(dither))
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PaletteQuantizer{TPixel}"/> class.
+        /// </summary>
+        /// <param name="palette">The color palette to use.</param>
+        /// <param name="diffuser">The error diffusion algorithm, if any, to apply to the output image</param>
+        public PaletteQuantizer(TPixel[] palette, IErrorDiffuser diffuser)
+        {
+            Guard.MustBeBetweenOrEqualTo(palette.Length, QuantizerConstants.MinColors, QuantizerConstants.MaxColors, nameof(palette));
+            this.palette = palette;
+            this.Diffuser = diffuser;
+        }
+
+        /// <inheritdoc/>
+        public IErrorDiffuser Diffuser { get; }
+
+        /// <inheritdoc/>
+        public IFrameQuantizer<TPixel1> CreateFrameQuantizer<TPixel1>(Configuration configuration)
+            where TPixel1 : struct, IPixel<TPixel1>
+        {
+            if (!typeof(TPixel).Equals(typeof(TPixel1)))
+            {
+                throw new InvalidOperationException("Generic method type must be the same as class type.");
```

Comment:
The error would only occur if someone did this which would compile because I'm having to hack the generic constraints. 

``` c#
var quantizer = new PaletteQuantizer<Rgba32>(palette)l
var frameQuantizer = quantizer.CreateFrameQuantizer<Rgb24>(configuration)
```



What I will do is mask the generic methods in that class with non-generic versions.

``` c#
var quantizer = new PaletteQuantizer<Rgba32>(palette)l
var frameQuantizer = quantizer.CreateFrameQuantizer(configuration)
```

The only way for them to cause the error now would be to explicitly cast...

``` c#
var quantizer = new PaletteQuantizer<Rgba32>(palette)l
var frameQuantizer = ((IQuantizer)quantizer).CreateFrameQuantizer<Rgb24>(configuration)
```
In which case the error message should be sufficient.